src/client/core/aad/authentication/authentication.service.ts

```diff
@@ -61,28 +61,42 @@ export class LogoutError extends SanitizedError {
  */
 export class AuthenticationService {
     public state: Observable<AuthenticationState>;
-    private _waitingForAuth = false;
-    private _authorizeQueue: AuthorizeQueueItem[] = [];
-    private _currentAuthorization: AuthorizeQueueItem | null = null;
+    private _authQueue = new AuthorizeQueue();
     private _state = new BehaviorSubject(AuthenticationState.None);
     private _logoutDeferred: Deferred<void> | null;
 
-    constructor(private app: BatchExplorerApplication, private config: AADConfig) {
+    constructor(
+        private app: BatchExplorerApplication,
+        private config: AADConfig,
+        private authProvider: AuthProvider
+    ) {
         this.state = this._state.asObservable();
     }
+
     /**
-     * Authorize the user.
-     * @param silent If set to true it will not ask the user for prompt. (i.e prompt=none for AD)
-     *      This means the request will fail if the user didn't give consent yet or it expired
+     * Authorize the user against the specified tenant.
      * @returns Observable with the successful AuthorizeResult.
-     *      If silent is true and the access fail the observable will return and error of type AuthorizeError
      */
-    public async authorize(tenantId: string, silent = false): Promise<AuthorizeResult> {
-        if (this._isAuthorizingTenant(tenantId)) {
-            return this._getTenantDeferred(tenantId).promise;
+    public async authorize(tenantId: string): Promise<AuthorizeResult> {
+        return this.authorizeResource(
+            tenantId, this.app.properties.azureEnvironment.arm
+        );
+    }
+
+    /**
+     * Authorizes a resource
+     *
+     * @param tenantId The tenant ID
+     * @param resourceURI The resource URI
+     * @returns a promise with AuthorizedResult
+     */
+    public async authorizeResource(tenantId: string, resourceURI: string):
```

Comment:
I don't see an await here, either. I didn't think you could use async by explicitly returning the promise object, but from the below I'm guessing I'm wrong about that.

Reply:
The `async` is extraneous because of the implementation (but it's not wrong, since the function returns a promise). But the other public `authorize*` methods are async so this makes it consistent and also allows for future modifications.